File: src/js/canvas_debug_draw.js

class CanvasDebugDraw {
    constructor(canvas) {
        this.canvas = canvas;
        this.context = canvas.getContext("2d")
    }


    draw_circle_impl(center, radius, color, axis){
        this.context.beginPath()
        this.context.arc(center[0], center[1], radius, 0, 2 * Math.PI, false)
        
        if(axis !== undefined)
        {
            this.context.fillStyle = `rgb(${color[0]},${color[1]},${color[2]})`
            this.context.fill()
        }
        else
        {
            this.context.strokeStyle = `rgb(${color[0]},${color[1]},${color[2]})`
            this.context.lineWidth = 1
            this.context.stroke()
        }
    }

    draw_segment_impl(points, color){
        this.context.beginPath();
        this.context.moveTo(points[0],points[1]);
        this.context.lineTo(points[2],points[3]);
        this.context.strokeStyle = `rgb(${color[0]},${color[1]},${color[2]})`;
        this.context.lineWidth = 1
        this.context.stroke()
    }

    draw_segment(p1,p2, color, lineWidth=1){
        this.context.beginPath();
        this.context.moveTo(p1[0],p1[1]);
        this.context.lineTo(p2[0],p2[1]);
        this.context.strokeStyle = `rgb(${color[0]},${color[1]},${color[2]})`;
        this.context.lineWidth = lineWidth
        this.context.stroke()
    }

    draw_poly_impl(polygon_length, polygon_points, color, solid, lineWidth=1){
        this.context.beginPath();

        this.context.moveTo(polygon_points[0],polygon_points[1]);
        for(let i=1; i<polygon_length; i++)
        {
            this.context.lineTo(polygon_points[2*i],polygon_points[2*i+1]);
        }
        this.context.lineTo(polygon_points[0],polygon_points[1]);
        this.context.closePath();
        if(solid)
        {
            this.context.fillStyle = `rgb(${color[0]},${color[1]},${color[2]})`;
            this.context.fill()
        }
        else
        {
            this.context.strokeStyle = `rgb(${color[0]},${color[1]},${color[2]})`;
            this.lineWidth = lineWidth
            this.context.stroke()
        }
    }


    draw_polygons(points, sizes, colors, solid){

        const n = sizes.length
        var offset = 0
        for (let i = 0; i < n; i++) {
            const polygon_length = sizes[i];
            const polygon_points = points.subarray(offset, 2*polygon_length)
            const polygon_color = colors.subarray(i*3, i*3 +3)
            this.draw_poly_impl(polygon_length,polygon_points,polygon_color, solid)
            offset += 2 * polygon_length;
        }
    }


    draw_solid_circles(centers, radii, axis, colors){
        const n = radii.length
        for (let i = 0; i < n; i++) {
            const center = centers.subarray(2*i, 2*i+2)
            const radius = radii[i]
            const color = colors.subarray(3*i, 3*i+3)
            const ax = axis.subarray(2*i, 2*i+2)
            this.draw_circle_impl(center, radius, color, ax)
        }
    }
    draw_circles(centers, radii, colors){
        const n = radii.length
        for (let i = 0; i < n; i++) {
            const center = centers.subarray(2*i, 2*i+2)
            const radius = radii[i]
            const color = colors.subarray(3*i, 3*i+3)
            this.draw_circle_impl(center, radius, color, undefined)
        }
    }
    draw_segments(points, colors){
        const n = points.length / 4
        for (let i = 0; i < n; i++) {
            const p = points.subarray(4*i, 4*i+4)
            const color = colors.subarray(3*i, 3*i+3)
            this.draw_segment_impl(p, color)
        }
    }
}

export {CanvasDebugDraw}

File: python/canvas_debug_draw.py
import pyjs
import b2d


class HtmlCanvasBatchDebugDraw(b2d.batch_debug_draw_cls(False, False, True)):
    def __init__(self, canvas, js_canvas_debug_draw, flags=None):
        super(HtmlCanvasBatchDebugDraw, self).__init__()
        self.canvas = canvas
        self.context = canvas.getContext("2d")
        self.js_canvas_debug_draw = js_canvas_debug_draw

        # what is drawn
        if True or flags is None:
            flags = [
                "shape",
                "joint",
                "particle",
            ]  # ,'aabb','pair','center_of_mass','particle']
        self.flags = flags
        self.clear_flags(
            ["shape", "joint", "aabb", "pair", "center_of_mass", "particle"]
        )
        for flag in flags:
            self.append_flags(flag)

    def _draw_solid_polygons(self, points, sizes, colors):
        
        js_points = pyjs.buffer_to_js_typed_array(points.ravel())
        js_sizes = pyjs.buffer_to_js_typed_array(sizes)
        js_colors = pyjs.buffer_to_js_typed_array(colors.ravel())

        self.js_canvas_debug_draw.draw_polygons(js_points, js_sizes, js_colors, True)

    def _draw_polygons(self, points, sizes, colors):
        
        js_points = pyjs.buffer_to_js_typed_array(points.ravel())
        js_sizes = pyjs.buffer_to_js_typed_array(sizes)
        js_colors = pyjs.buffer_to_js_typed_array(colors.ravel())

        self.js_canvas_debug_draw.draw_polygons(js_points, js_sizes, js_colors, False)

    def _draw_solid_circles(self, centers, radii, axis, colors):
        
        js_centers = pyjs.buffer_to_js_typed_array(centers.ravel())
        js_radii = pyjs.buffer_to_js_typed_array(radii)
        js_axis = pyjs.buffer_to_js_typed_array(axis.ravel())
        js_colors = pyjs.buffer_to_js_typed_array(colors.ravel())

        self.js_canvas_debug_draw.draw_solid_circles(
            js_centers, js_radii, js_axis, js_colors
        )

    def _draw_circles(self, centers, radii, colors):
        
        js_centers = pyjs.buffer_to_js_typed_array(centers.ravel())
        js_radii = pyjs.buffer_to_js_typed_array(radii)
        js_colors = pyjs.buffer_to_js_typed_array(colors.ravel())

        self.js_canvas_debug_draw.draw_circles(js_centers, js_radii, js_colors)

    def _draw_points(self, centers, sizes, colors):
        return

    def _draw_segments(self, points, colors):
        js_points = pyjs.buffer_to_js_typed_array(points.ravel())
        js_colors = pyjs.buffer_to_js_typed_array(colors.ravel())
        self.js_canvas_debug_draw.draw_segments(js_points, js_colors)

    def _draw_particles(self, centers, radius, colors=None):
        js_centers = pyjs.buffer_to_js_typed_array(centers.ravel())
        if colors is not None:
            colors = numpy.require(colors, dtype='uint8')

            js_colors = pyjs.buffer_to_js_typed_array(colors.ravel())
            self.js_canvas_debug_draw.draw_particles(js_centers, radius, js_colors)
        else:
            self.js_canvas_debug_draw.draw_particles(js_centers, radius)

    # non-batch api
    def draw_solid_circle(self, center, radius, axis, color):
        
        # print("draw_solid_circle")
        center = self.world_to_screen(center)
        radius = self.world_to_screen_scale(radius)
        center = numpy.require(center,dtype='uint32')
        axis = numpy.require(axis, dtype='float32')
        color = (numpy.require(color) * 255.0).astype("uint8")

        js_center = pyjs.buffer_to_js_typed_array(center.ravel())
        js_axis = pyjs.buffer_to_js_typed_array(axis.ravel())
        js_color = pyjs.buffer_to_js_typed_array(color.ravel())

        self.js_canvas_debug_draw.draw_circle_impl(js_center, radius, js_color, js_axis)

    def draw_circle(self, center, radius, color, line_width=1):
        
        center = self.world_to_screen(center)
        radius = self.world_to_screen_scale(radius)
        center = numpy.require(center, dtype='uint32')
        color = (numpy.require(color) * 255.0).astype("uint8")

        js_center = pyjs.buffer_to_js_typed_array(center.ravel())
        js_color = pyjs.buffer_to_js_typed_array(color.ravel())

        self.js_canvas_debug_draw.draw_circle_impl(js_center, radius, js_color)

    def draw_segment(self, p1, p2, color, line_width=1):
        p1 = pyjs.buffer_to_js_typed_array(numpy.require(self.world_to_screen(p1)))
        p2 = pyjs.buffer_to_js_typed_array(numpy.require(self.world_to_screen(p2)))
        color = (numpy.require(color) * 255.0).astype("uint8")
        js_color = pyjs.buffer_to_js_typed_array(color.ravel())
        self.js_canvas_debug_draw.draw_segment(
            p1, p2, js_color, self.world_to_screen_scale(line_width)
        )

    def draw_polygon(self, vertices, color, line_width=1):
        for i, ver in enumerate(vertices):
            vertices[i] = self.world_to_screen(ver)
        color = (numpy.require(color) * 255.0).astype("uint8")

        js_points = pyjs.buffer_to_js_typed_array(vertices.ravel())
        js_color = pyjs.buffer_to_js_typed_array(color.ravel())

        self.js_canvas_debug_draw.draw_poly_impl(
            int(vertices.shape[0]),
            js_points,
            js_color,
            False,
            self.world_to_screen_scale(line_width),
        )

    def draw_solid_polygon(self, vertices, color):
        for i, ver in enumerate(vertices):
            vertices[i] = self.world_to_screen(ver)
        color = (numpy.require(color) * 255.0).astype("uint8")

        js_points = pyjs.buffer_to_js_typed_array(vertices.ravel())
        js_color = pyjs.buffer_to_js_typed_array(color.ravel())

        self.js_canvas_debug_draw.draw_poly_impl(
            int(vertices.shape[0]), js_points, js_color, True
        )
